src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -60,11 +62,43 @@ private void Initialize()
         {
             lock (_lock)
             {
-                _changeToken = new CompositeChangeToken(_dataSources.Select(d => d.ChangeToken).ToArray());
                 _endpoints = _dataSources.SelectMany(d => d.Endpoints).ToArray();
 
-                _changeToken.RegisterChangeCallback((state) => Initialize(), null);
+                foreach (var dataSource in _dataSources)
+                {
+                    Extensions.Primitives.ChangeToken.OnChange(
+                        () => dataSource.ChangeToken,
+                        () => HandleChange());
+                }
             }
         }
+
+        private void HandleChange()
+        {
+            // Refresh the endpoints from datasource so that callbacks can get the latest endpoints
+            Initialize();
```

Comment:
This whole method should also take the lock.

Reply:
Also, add a comment explaining that you are purposely updating the endpoints **before** triggering the public change token. This is important because it avoids the need for a lock to access the endpoints.

Reply:
I would suggest that you don't call `Initialize` here - that will create redundant change token registrations. Just copy-paste line 65